packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -67,15 +123,106 @@ class Constants : public GadgetT
 
     void generate_r1cs_constraints()
     {
-        pb.add_r1cs_constraint(ConstraintT(zero, FieldT::one(), FieldT::zero()), ".zero");
-        pb.add_r1cs_constraint(ConstraintT(one, FieldT::one(), FieldT::one()), ".one");
+        pb.add_r1cs_constraint(ConstraintT(_0, FieldT::one(), FieldT::zero()), ".zero");
```

Comment:
In theory not, because a variable can be a linear combination of multiple variables. So you can e.g. always use `3*a + 2*b + 1*c + d` as the value in a constraint. This is heavily used in some places (like for poseidon) because it can greatly optimize the constraints. In normal cases however it's a bit of pain to work with and it's easier to just use normal variables. We need to write these by hand so it's not easy to make use of it and in normal cases it can't really be used so save constraints.

Reply:
Not really, but a bit. :)

By using a `linear_combination<FieldT>` instead of a `pb_variable<FieldT>` you can e.g. add constants to a single "variable". So for example if you have `a` and `b` you can combine them in a single linear combination like `2*a + 3*b+ 5` and use this "single variable" in a constraint to save constraints (note that `a*b` cannot be done like this). So in some cases where these constant variables are used this could be used instead, instead of needing to use a constraint that does `3*a`. But this is rarely the case.

What I'm doing with these constants here is just putting them in a `pb_variable<FieldT>` so I can pass them around in all the gadgets without having to add special logic and make everything a `linear_combination<FieldT>` or manually substitute before they are actually used in a constraint. So in some cases this may waste a constraint because the multiplication/addition could be handled by a `linear_combination`, but you can see how this can get complicated fast for normal logic to actually make use of this. Unless in very specific circumstances like with Poseidon it doesn't make sense to do this (unless of course your circuit compiler could do it for you, but I don't think that's the case even in Bellman).

A good example how this is used for Poseidon which has these long linear combinations which greatly reduces the number of constraints because Poseidon was designed to specifically be able to make use of this: https://github.com/Loopring/ethsnarks/blob/f8ae6afdf544f0795003a715880336b364d0e2f5/src/gadgets/poseidon.hpp#L191